### DISCUSSION TITLE
Standardize rdk location across different screen sizes/resolutions

### DISCUSSION THREAD

**Discussion by wsr017** [jspsych/jsPsych#1122] [Q&A]
Hello!

I am trying to put a rdk field in a very specific vertical location on the screen, but it appears higher or lower depending on which computer is running the program. I have `aperture_center_y` defined as (`window.screen.height` * a fraction that doesn't change). I'd love some advice for how to standardize the location of the dot field on the screen -- for this particular program it needs to be extremely precise for any computer. 

For clarification, the equator of the dot field is supposed to be aligned with the eyes of a previously shown face. This is shown by my added red arrow in this overlay from a laptop screen (16 inch, 3072 x 1920):
![new small line](https://user-images.githubusercontent.com/56043266/96176130-ca094380-0ee0-11eb-96e5-d107080dba03.png)

however when I run the same program on a different, larger monitor (23.5 inch, 2560 x 1440), the dot field has moved down:
![new big line](https://user-images.githubusercontent.com/56043266/96176162-d1c8e800-0ee0-11eb-85ce-439779217502.png)

The face image is not changing locations, so to my knowledge the issue must be with the rdk field location. 

Thanks so much for any help!

**Comment by giuconte** [comment]
I'm experiencing something similar, I'm replying here since the two issues might be related.

when running in a normal window the stimuli are centered, while on a full-screen window they are slightly moved upward from the center. I do not have the same behavior with other plugins which are usually centered on both full-screen and reduce-screen modality.
any idea how to achieve this?

thanks

Giulio

**Comment by wsr017** [comment]
To give some more info, here's the relevant code for rdk positioning:

```
var fs_width = window.screen.width;
var fs_height = window.screen.height;

var face_tree_image_width=fs_width*0.4;
var dot_aperture_width=fs_width*0.1376; //these multipliers position it correctly on my 16 inch screen

var dot_aperture_center_y = fs_height*0.407;
var dot_aperture_center_x = fs_width*0.447;

```

The facial stimulus is presented like this:

```
var facial_stimuli = {
        type: 'image-button-response',
        stimulus: jsPsych.timelineVariable('faceType'),
        stimulus_width: face_tree_image_width,
        choices: [],
        trial_duration: faceStimDuration
      }
```

and the RDK is presented like this:

```
var dot_motion = {
    type: "rdk",
    number_of_dots: n_dots,
    coherence: jsPsych.timelineVariable('coher_level'),
    move_distance: dot_speed, 
    reinsertType: 1, // 1 = random, 2 = continuous Square
    aperture_type: 3, // 3 = square
    RDK_type: 3, // same and random direction
    background_color: '#d2d2d2',
    dot_color: 'black',
    aperture_width: dot_aperture_width,
    aperture_height: dot_aperture_width,
    aperture_center_y: dot_aperture_center_y,
    aperture_center_x: dot_aperture_center_x,
    response_ends_trial: true,
    trial_duration: dotsInterval,
    response_ends_trial: true,
    coherent_direction: jsPsych.timelineVariable('dotDir'),
    oppositeCoherence: 0,
    dot_life: dot_lifetime,
    dot_radius: dot_radius,
    data: {congruence: jsPsych.timelineVariable('cong')}
}
```



I've adjusted the rdk location a little bit since the original post, here are new images showing the difference. 
16 inch screen, blue lines are the same length, eyes are in the middle:
<img width="1007" alt="Screen Shot 2020-10-27 at 12 11 02 PM" src="https://user-images.githubusercontent.com/56043266/97350050-7c2ffc00-184d-11eb-8f0a-8d80b8c4b5bd.png">

23.5 inch screen where the eyes are now above the equator of the rdk (the blue lines are the same length).  
<img width="1212" alt="Screen Shot 2020-10-27 at 12 17 04 PM" src="https://user-images.githubusercontent.com/56043266/97350929-548d6380-184e-11eb-8eba-85d7ed23c341.png">


I realize this seems a little trivial, but I wondered if there was a way to standardize it. Thanks!

**Comment by becky-gilbert** [comment]
Hi @wsr017, sorry for the delay getting back to you. I think the problem might be that you're using `window.screen.height`/`window.screen.width`, whereas the RDK plugin positions the canvas using `window.innerHeight`/`window.innerWidth`. So I changed that in the code below, and positioned the aperture center using half of the innerHeight and innerWidth values. This is a solution that @vrsivananda suggested in [this thread](https://github.com/jspsych/jsPsych/pull/681). And I followed Sivananda's suggestion to make all the parameters that rely on the window size into functions, so that the trials use the current browser window size when the trial runs (just in case the window size was changed sometime between when the page first loaded and the start of the trial). 

Also, I added some CSS to the canvas element to get rid of the scroll bars, in case that hasn't been fixed in your version of the plugin (see [here](https://github.com/jspsych/jsPsych/issues/881)). The other parameters that I changed were just for more convenient testing.

Here's my code:
```html
    <style>
        body {
            background-color: #d2d2d2;
        }
        canvas {
            position: absolute;
            top: 0;
            left: 0;
        }
    </style>
```
```javascript
    var facial_stimuli = {
        type: 'image-button-response',
        stimulus: "eyesopen_looking_right.jpg",
        stimulus_width: function() {
            return window.innerWidth*0.4;
        },
        choices: [],
        trial_duration: 5000
    };

    var dot_motion = {
        type: "rdk",
        number_of_dots: 300,
        coherence: 1,
        move_distance: 1, 
        reinsertType: 1, // 1 = random, 2 = continuous Square
        aperture_type: 3, // 3 = square
        RDK_type: 3, // same and random direction
        background_color: '#d2d2d2',
        dot_color: 'black',
        aperture_width: function() {
            return window.innerWidth*0.1376;
        },
        aperture_height: function() {
            return window.innerWidth*0.1376;
        },
        aperture_center_y:  function(){
		    return window.innerHeight/2;
	    },
        aperture_center_x:  function(){
		    return window.innerWidth/2;
        },
        response_ends_trial: true,
        trial_duration: 5000,
        correct_choice: 'a'
    };
```
I'm not able to test this on different monitors, but I tried a few different window sizes and it seems to work. And I've only eyeballed the stimuli positioning, rather than testing it with any kind of precision. Can you please let me know if it works for you?
